velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -703,9 +780,107 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+// floating-point columns moved to the end and rows in ascending order.
+bool tryBucketMaterializedRows(
+    const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right,
+    std::vector<MaterializedRow>& leftReordered,
+    std::vector<MaterializedRow>& rightReordered) {
+  if (left.empty() || right.empty() ||
+      left.begin()->size() != right.begin()->size()) {
+    return false;
+  }
+
+  std::vector<velox::column_index_t> floatColumnIndices;
+  auto& leftRow = *left.begin();
+  auto& rightRow = *right.begin();
+  auto width = leftRow.size();
+  for (auto i = 0; i < width; ++i) {
+    if (leftRow[i].kind() != rightRow[i].kind()) {
+      return false;
+    }
+    if ((leftRow[i].kind() == TypeKind::REAL ||
+         leftRow[i].kind() == TypeKind::DOUBLE)) {
+      floatColumnIndices.push_back(i);
+    }
+  }
+
+  auto moveFloatColumnsToEnd =
+      [](const std::vector<velox::column_index_t>& floatColumnIndices,
+         auto width,
+         std::vector<MaterializedRow>& reorderedResults) {
+        for (auto i = 1; i < reorderedResults.size(); ++i) {
+          velox::column_index_t lastNonFloatColumn = width - 1;
+          for (auto j = (int32_t)floatColumnIndices.size() - 1; j >= 0; --j) {
+            if (floatColumnIndices[j] != lastNonFloatColumn) {
+              std::swap(
+                  reorderedResults[i][floatColumnIndices[j]],
+                  reorderedResults[i][lastNonFloatColumn]);
+            }
+            --lastNonFloatColumn;
+          }
+        }
+      };
+  moveFloatColumnsToEnd(floatColumnIndices, width, leftReordered);
+  moveFloatColumnsToEnd(floatColumnIndices, width, rightReordered);
+
+  std::sort(leftReordered.begin(), leftReordered.end());
+  std::sort(rightReordered.begin(), rightReordered.end());
+
+  // Check that every bucket grouped by the first width - 1 columns has only
+  // one row.
+  auto checkBucketSize = [](const std::vector<MaterializedRow>& groupedBuckets,
+                            int bucketWidth,
+                            bool& oversized) {
+    for (auto i = 1; i < groupedBuckets.size(); ++i) {
+      oversized = true;
+      for (auto j = 0; j < bucketWidth; ++j) {
+        if (groupedBuckets[i][j] != groupedBuckets[i - 1][j]) {
+          oversized = false;
+          break;
+        }
+      }
+      if (oversized) {
+        break;
+      }
+    }
+  };
+  bool leftOversized = false;
+  bool rightOversized = false;
+  width = width - floatColumnIndices.size();
+  checkBucketSize(leftReordered, width, leftOversized);
+  checkBucketSize(rightReordered, width, rightOversized);
+
+  return !leftOversized || !rightOversized;
+}
+
+static bool compareMaterializedRows(
+    const std::vector<MaterializedRow>& leftReordered,
+    const std::vector<MaterializedRow>& rightReordered) {
+  auto width = leftReordered[0].size();
+  auto length = leftReordered.size();
+
+  // Compare row-by-row with epsilon.
```

Comment:
I thought about this, but was wondering whether there will be rows with different width in one MaterializedRowMultiset in practice, since it's either returned from DuckDB execution or converted from a RowVector. I can add this check for every row.